File: param_gauss_recon/Module/pcd_sampler.py
import os
import numpy as np
import open3d as o3d
from typing import Union

from param_gauss_recon.Method.path import createFileFolder
from param_gauss_recon.Method.pcd import downSample, getPointCloud


class PcdSampler(object):
    def __init__(self) -> None:
        return

    def toFPSPcd(
        self, pcd: o3d.geometry.PointCloud, sample_point_num: int
    ) -> Union[o3d.geometry.PointCloud, None]:
        if np.asarray(pcd.points).shape[0] < sample_point_num:
            return pcd

        down_sample_pcd = downSample(pcd, sample_point_num)

        if down_sample_pcd is None:
            print("[ERROR][PcdSampler::toFPSPcd]")
            print("\t downSample failed!")
            return None

        return down_sample_pcd

    def toFPSPoints(
        self, points: np.ndarray, sample_point_num: int
    ) -> Union[np.ndarray, None]:
        if points.shape[0] < sample_point_num:
            return points

        pcd = getPointCloud(points)

        down_sample_pcd = self.toFPSPcd(pcd, sample_point_num)

        if down_sample_pcd is None:
            print("[ERROR][PcdSampler::toFPSPoints]")
            print("\t downSample failed!")
            return None

        return down_sample_pcd

    def toFPSPcdFile(
        self, pcd_file_path: str, sample_point_num: int, save_pcd_file_path: str
    ) -> bool:
        if not os.path.exists(pcd_file_path):
            print("[ERROR][PcdSampler::toFPSPcdFile]")
            print("\t pcd file not exist!")
            print("\t pcd_file_path:", pcd_file_path)
            return False

        pcd = o3d.io.read_point_cloud(pcd_file_path)

        down_sample_pcd = self.toFPSPcd(pcd, sample_point_num)

        if down_sample_pcd is None:
            print("[ERROR][PcdSampler::toFPSPcdFile]")
            print("\t downSample failed!")
            return False

        createFileFolder(save_pcd_file_path)

        o3d.io.write_point_cloud(save_pcd_file_path, down_sample_pcd, write_ascii=True)
        return True

File: src/pcd_sampler.cpp
#include "pcd_sampler.h"
#include "pcd.h"
#include <filesystem>

std::shared_ptr<open3d::geometry::PointCloud> PcdSampler::toFPSPcd(
    std::shared_ptr<open3d::geometry::PointCloud> &pcd,
    const int &sample_point_num){
  const int pcd_point_num = pcd->points_.size();
  if (pcd_point_num < sample_point_num){
    std::cout << "[WARN][PcdSampler::toFPSPcd]" << std::endl;
    std::cout << "\t pcd point num = " << pcd_point_num << " , < sample point num = " << sample_point_num << " !" << std::endl;
    std::cout << "\t will directly return source point cloud!" << std::endl;

    return pcd;
  }

  std::shared_ptr<open3d::geometry::PointCloud> down_sample_pcd = downSample(pcd, sample_point_num);

  return down_sample_pcd;
}

const bool PcdSampler::toFPSPcdFile(
    const std::string &pcd_file_path,
    const int &sample_point_num,
    const std::string &save_pcd_file_path){
  if (!std::filesystem::exists(pcd_file_path)){
    std::cout << "[ERROR][PcdSampler::toFPSPcdFile]" << std::endl;
    std::cout << "\t pcd file not exist!" << std::endl;
    std::cout << "\t pcd_file_path : " << pcd_file_path << std::endl;
    return false;
  }

  std::shared_ptr<open3d::geometry::PointCloud> pcd = std::make_shared<open3d::geometry::PointCloud>();
  if (!open3d::io::ReadPointCloud(pcd_file_path, *pcd)){
    std::cout << "[ERROR][PcdSampler::toFPSPcdFile]" << std::endl;
    std::cout << "\t ReadPointCloud failed!" << std::endl;
    std::cout << "\t pcd_file_path : " << pcd_file_path << std::endl;
    return false;
  }

  const std::string save_pcd_folder_path = std::filesystem::path(save_pcd_file_path).parent_path();

  std::shared_ptr<open3d::geometry::PointCloud> down_sample_pcd = toFPSPcd(pcd, sample_point_num);

  if (!std::filesystem::exists(save_pcd_folder_path)){
    std::filesystem::create_directories(save_pcd_folder_path);
  }

  if (!open3d::io::WritePointCloud(save_pcd_file_path, *down_sample_pcd, open3d::io::WritePointCloudOption("auto",
          open3d::io::WritePointCloudOption::IsAscii::Ascii))){
    std::cout << "[ERROR][PcdSampler::toFPSPcdFile]" << std::endl;
    std::cout << "\t WritePointCloud failed!" << std::endl;
    std::cout << "\t save_pcd_file_path : " << save_pcd_file_path << std::endl;
    return false;
  }

  return true;
}
